river/ensemble/adaptive_random_forest.py

```diff
@@ -0,0 +1,824 @@
+        self.max_depth = max_depth
+
+    def _multiclass(self):
+        return True
+
+    def predict_proba_one(self, x: dict) -> typing.Dict[base.typing.ClfTarget, float]:
+
+        y_pred = collections.Counter()
+
+        if not self.models:
+            self._init_ensemble(features=list(x.keys()))
+            return y_pred
+
+        for model in self.models:
+            y_proba_temp = model.predict_proba_one(x)
+            metric_value = model.metric.get()
+            if not self.disable_weighted_vote and metric_value > 0.:
+                y_proba_temp = {k: val * metric_value for k, val in y_proba_temp.items()}
+            y_pred.update(y_proba_temp)
+
+        total = sum(y_pred.values())
+        if total > 0:
+            return {label: proba / total for label, proba in y_pred.items()}
+        return y_pred
+
+    def _get_base_model(self, seed: int):
+        return ARFHoeffdingTreeClassifier(
+                    max_size=self.max_size,
+                    memory_estimate_period=self.memory_estimate_period,
+                    grace_period=self.grace_period,
+                    split_criterion=self.split_criterion,
+                    split_confidence=self.split_confidence,
+                    tie_threshold=self.tie_threshold,
+                    binary_split=self.binary_split,
+                    stop_mem_management=self.stop_mem_management,
+                    remove_poor_attrs=self.remove_poor_attrs,
+                    merit_preprune=self.merit_preprune,
+                    leaf_prediction=self.leaf_prediction,
+                    nb_threshold=self.nb_threshold,
+                    nominal_attributes=self.nominal_attributes,
+                    max_features=self.max_features,
+                    max_depth=self.max_depth,
+                    seed=seed
+                )
+
+
+class AdaptiveRandomForestRegressor(BaseForest, base.Regressor):
+    r"""Adaptive Random Forest regressor.
+
+    The 3 most important aspects of Adaptive Random Forest [^1] are:
+    1. inducing diversity through re-sampling;
+    2. inducing diversity through randomly selecting subsets of features for
+       node splits
+    3. drift detectors per base tree, which cause selective resets in response
+       to drifts.
+
+    Notice that this implementation is slightly different from the original
+    algorithm proposed in [^2]. The `HoeffdingTreeRegressor` is used as base
+    learner, instead of `FIMT-DD`. It also adds a new strategy to monitor the
+    incoming data and check for concept drifts. The monitored data (either the
+    trees' errors or their predictions) are centered and scaled (z-score
+    normalization) to have zero mean and unit standard deviation. Transformed
+    values are then again normalized in the [0, 1] range to fulfil ADWIN's
+    requirements. We assume that the data subjected to the z-score
+    normalization lies within the interval of the mean $\pm3\sigma$, as it
+    occurs in normal distributions.
```

Comment:
Something along these lines (to reflect the new changes):

> The monitored data (either the trees' errors or their predictions) are normalized in the [0, 1] range to fulfil ADWIN's requirements. We assume that the data subjected to the normalization, e.g., the monitored error, follows a normal distribution, and thus, lies within the interval of the mean $\pm3\sigma$.

Reply:
Updated the documentation in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

Reply:
Updated again in b628832e1062fda4cdbb7ae1c1948d4d6212bd90 to account for the updates to the drift detection strategy.